### ISSUE TITLE
ProofGeneral freezes when there is a build error in a dependency file

### ISSUE THREAD

**Issue by RalfJung** [ProofGeneral/PG#572]
With "build dependencies" mode enabled, usually PG shows error messages when a dependency fails to build. Recently, it started to just hang instead, so that I have to kill emacs entirely.

My Coq/PG-related settings are:
```
 '(coq-compile-before-require t)
 '(coq-compile-parallel-in-background t)
 '(coq-compile-quick 'no-quick)
 '(coq-compile-vos 'vos)
 '(coq-one-command-per-line nil)
 '(proof-follow-mode 'followdown)
 '(proof-next-command-insert-space nil)
 '(proof-splash-enable nil)
 '(proof-three-window-enable t)
```

**Comment by hendriktews** [comment]
> to just hang instead, so that I have to kill emacs entirely.

C-g does not help?

What kind of errors do Proof General cause to hang? Just syntax
errors or errors in proofs? Or circular dependencies?

Needless to say, without something to reproduce, it's going to be
difficult...

Can you share some more details? How often does it strike? Is
there any particular usage pattern?

You can try to set coq-compile-keep-going to nil in order to kill
everything when the first error is recognized.

**Comment by hendriktews** [comment]
The errors that cause this - to they case coqdep or coqc to fail?

**Comment by hendriktews** [comment]
Do you assert all require commands together? Or do you extend the
queue region with additional require commands while the first
background compilation is still running?

The file that contains the error, is it directly required in the
asserted region? Or is it only indirectly required by some other
file?

**Comment by RalfJung** [comment]
> You can try to set coq-compile-keep-going to nil in order to kill
everything when the first error is recognized.

That actually makes a difference! With that setting I now get the expected
```
-*- mode: compilation; -*-

coqc -vos -w -ssr-search-moved -w -deprecated-hint-without-locality -w +deprecated-tactic-notation -w -notation-overridden,-redundant-canonical-projection -Q /home/r/Dokumente/Unisachen/perennial/coq/src Perennial -Q /home/r/Dokumente/Unisachen/perennial/coq/external/stdpp/theories stdpp -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris iris -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris_deprecated iris.deprecated -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris_staging iris.staging -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris_heap_lang iris.heap_lang -Q /home/r/Dokumente/Unisachen/perennial/coq/external/coqutil/src/coqutil coqutil -Q /home/r/Dokumente/Unisachen/perennial/coq/external/Goose Goose -Q /home/r/Dokumente/Unisachen/perennial/coq/external/record-update/src RecordUpdate -Q /home/r/Dokumente/Unisachen/perennial/coq/external/coq-tactical/src Tactical -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris-named-props/src iris_named_props /home/r/Dokumente/Unisachen/perennial/coq/src/program_logic/atomic.v
File "/home/r/Dokumente/Unisachen/perennial/coq/src/program_logic/atomic.v", line 36, characters 0-1:
Error: Syntax error: illegal begin of vernac.
```

**Comment by cpitclaudel** [comment]
Can you `pkill -SIGUSR2 emacs` while it's frozen?

**Comment by RalfJung** [comment]
Okay that did a thing, a buffer popped up in emacs saying
```
Debugger entered--Lisp error: (quit)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  coq-par-ongoing-compilation(coq-compile-job-symbol)
  #f(compiled-function (anc-job not-used) #<bytecode 0x158f843c593d>)(coq-compile-job-symbol t)
  maphash(#f(compiled-function (anc-job not-used) #<bytecode 0x158f843c593d>) #<hash-table eql 172/264 0x158f8458050d>)
  coq-par-unlock-job-ancestors-on-error(coq-compile-job-symbol)
  coq-par-mark-job-failing(coq-compile-job-symbol)
  coq-par-mark-job-failing(coq-compile-job-symbol)
  coq-par-mark-job-failing(coq-compile-job-symbol)
  coq-par-coqc-continuation(#<process pro-455> 1)
  coq-par-process-sentinel(#<process pro-455> "exited abnormally with code 1\n")
```
Also in the "coq-compile-response" buffer I can see
```
-*- mode: compilation; -*-

coqc -vos -w -ssr-search-moved -w -deprecated-hint-without-locality -w +deprecated-tactic-notation -w -notation-overridden,-redundant-canonical-projection -Q /home/r/Dokumente/Unisachen/perennial/coq/src Perennial -Q /home/r/Dokumente/Unisachen/perennial/coq/external/stdpp/theories stdpp -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris iris -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris_deprecated iris.deprecated -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris_staging iris.staging -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris/iris_heap_lang iris.heap_lang -Q /home/r/Dokumente/Unisachen/perennial/coq/external/coqutil/src/coqutil coqutil -Q /home/r/Dokumente/Unisachen/perennial/coq/external/Goose Goose -Q /home/r/Dokumente/Unisachen/perennial/coq/external/record-update/src RecordUpdate -Q /home/r/Dokumente/Unisachen/perennial/coq/external/coq-tactical/src Tactical -Q /home/r/Dokumente/Unisachen/perennial/coq/external/iris-named-props/src iris_named_props /home/r/Dokumente/Unisachen/perennial/coq/src/program_logic/atomic.v
File "/home/r/Dokumente/Unisachen/perennial/coq/src/program_logic/atomic.v", line 36, characters 0-1:
Error: Syntax error: illegal begin of vernac.
```

**Comment by hendriktews** [comment]
I can reproduce the hanging PG, however, for me C-g makes emacs alive again and shows the error. No need to kill emacs here.

**Comment by RalfJung** [comment]
I tried this again, and after doing C-g I have to wait another 5-10s during which emacs is still entirely unresponsive (not even the cursor is blinking), but then it indeed comes back -- I don't think I tried this particular shortcut, and if I did I guess I was not patient enough.

**Comment by hendriktews** [comment]
There is another algorithm that accidentally has exponential behavior, causing emacs to hang.

**Comment by hendriktews** [comment]
The fix for this issue is ready in PR #577.
@RalfJung : Are you interested in trying this PR before we merge?
I checked manually that the reported error is fixed. Besides, the PR also adds a number of tests that check that in case of error unlocking of the queue region and of ancestor files works as expected.
